// devcert-with-localhost/dist/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const path = require("path");
const mkdirp = require("mkdirp");
exports.isMac = process.platform === 'darwin';
exports.isLinux = process.platform === 'linux';
exports.isWindows = process.platform === 'win32';
if (exports.isWindows && process.env.LOCALAPPDATA) {
    exports.configDir = path.join(process.env.LOCALAPPDATA, 'devcert', 'config');
}
else {
    let uid = process.getuid && process.getuid();
    let userHome = (exports.isLinux && uid === 0) ? path.resolve('/usr/local/share') : require('os').homedir();
    exports.configDir = path.join(userHome, '.config', 'devcert');
}
exports.configPath = path.join.bind(path, exports.configDir);
exports.opensslConfTemplate = path.join(__dirname, '..', 'openssl.conf');
exports.opensslConfPath = exports.configPath('openssl.conf');
exports.rootKeyPath = exports.configPath('devcert-ca-root.key');
exports.rootCertPath = exports.configPath('devcert-ca-root.crt');
exports.caCertsDir = exports.configPath('certs');
mkdirp.sync(exports.configDir);
mkdirp.sync(exports.caCertsDir);
