// devcert-with-localhost/dist/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const fs_1 = require("fs");
const child_process_1 = require("child_process");
const createDebug = require("debug");
const path = require("path");
const constants_1 = require("./constants");
const debug = createDebug('devcert');
function openssl(cmd) {
    return run(`openssl ${cmd}`, {
        stdio: 'ignore',
        env: Object.assign({
            RANDFILE: path.join(constants_1.configPath('.rnd'))
        }, process.env)
    });
}
exports.openssl = openssl;
function run(cmd, options = {}) {
    debug(`exec: \`${cmd}\``);
    return child_process_1.execSync(cmd, options);
}
exports.run = run;
function waitForUser() {
    return new Promise((resolve) => {
        process.stdin.resume();
        process.stdin.on('data', resolve);
    });
}
exports.waitForUser = waitForUser;
// Generate a cryptographic key, used to sign certificates or certificate signing requests.
function generateKey(filename) {
    debug(`generateKey: ${filename}`);
    openssl(`genrsa -out ${filename} 2048`);
    fs_1.chmodSync(filename, 400);
}
exports.generateKey = generateKey;
